webapp/channels/src/components/permalink_view/actions.ts

```diff
@@ -88,25 +89,48 @@ export function focusPost(postId: string, returnTo = '', currentUserId: string,
         if (privateChannelJoinPromptVisible) {
             return;
         }
-        const {data} = await dispatch(getPostThread(postId));
 
-        if (!data) {
+        let postInfo;
+        try {
+            postInfo = await Client4.getPostInfo(postId);
+        } catch (e) {
             getHistory().replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
             return;
         }
 
-        if (data.first_inaccessible_post_time) {
+        const state = getState();
+        const currentTeam = getCurrentTeam(state);
+
+        if (!postInfo.has_joined_channel) {
+            // Prompt system admin before joining the private channel
+            const user = getCurrentUser(state);
+            if (postInfo.channel_type === Constants.PRIVATE_CHANNEL && isSystemAdmin(user.roles)) {
+                privateChannelJoinPromptVisible = true;
+                const joinPromptResult = await dispatch(joinPrivateChannelPrompt(currentTeam, postInfo.channel_display_name));
+                privateChannelJoinPromptVisible = false;
+                if ('data' in joinPromptResult && !joinPromptResult.data.join) {
+                    return;
+                }
+            }
+            await dispatch(joinChannel(currentUserId, '', postInfo.channel_id));
+        }
+
+        const {data: threadData} = await dispatch(getPostThread(postId));
+
+        if (!threadData) {
+            getHistory().replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
+            return;
+        }
+
+        if (threadData.first_inaccessible_post_time) {
             getHistory().replace(`/error?type=${ErrorPageTypes.CLOUD_ARCHIVED}&returnTo=${returnTo}`);
             return;
         }
 
-        const state = getState();
         const isCollapsed = isCollapsedThreadsEnabled(state);
 
-        const channelId = data.posts[data.order[0]].channel_id;
-        let channel = state.entities.channels.channels[channelId];
-        const currentTeam = getCurrentTeam(state);
-        const teamId = currentTeam.id;
+        const channelId = threadData.posts[threadData.order[0]].channel_id;
+        let channel = getChannelFromRedux(state, channelId);
```

Comment:
That is exactly what it is doing (or at least what it should do). `getChannelFromRedux` is a selector that takes some information of the state. In this case the implementation is not much more than just what you said. Having this in a centralized function instead of calling just `state.entities.channels.channels[channelId]` is to make sure if any change is made to the state, it is only done in one place (in the implementation of `getChannelFromRedux`).

The name though is a bit unfortunate. It is called like that in this context because you can get the channel from two different places: the server or the local state (redux). So... since we need both in this file, we named the one to get the channel from the local state `getChannelFromRedux`.

Reply:
In other places, we've resolved that conflict by referring to the selectors as `selectChannel`, but that one might not work here. I've always wanted to rename the actions to `fetchChannel` to avoid any of that

Reply:
Thanks for explaining!